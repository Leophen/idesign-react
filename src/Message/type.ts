export interface MessageProps {
  /**
  * 全局提示类型
  * @default info
  */
  type?: 'info' | 'success' | 'warning' | 'error';
  /**
   * 全局提示内容
   */
  content?: string;
  /**
   * 消息显示时长，单位：秒。值为 0 表示永久显示
   * @default 3
   */
  duration?: number;
  /**
  * 全局提示位置
  * @default top
  */
  placement?: 'top' | 'bottom';
}

export interface MessageConfig extends MessageProps { }
