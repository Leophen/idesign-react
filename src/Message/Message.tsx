import React from 'react';
import './index.scss';
import { MessageProps } from './type';
import Icon from '../Icon';

const Message: React.FC<MessageProps> & {
  success?: any;
  warning?: any;
  error?: any;
  info?: any;
  closeAll?: any;
} = (props) => {
  const {
    type = 'info',
    content = '',
    ...restProps
  } = props;

  return (
    <div className="i-message" {...restProps}>
      <Icon
        name={
          {
            info: 'TipInfo',
            success: 'TipCheckFill',
            warning: 'TipWarningFill',
            error: 'TipCloseFill',
          }[type]
        }
        size={20}
      />
      {content}
    </div>
  );
};

Message.displayName = 'Message';

export default Message;
