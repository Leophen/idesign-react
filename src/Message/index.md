---
nav:
  title: 组件
  path: /components
group:
  title: 弹框提示组件
  order: 5
order: 2
---

# Message 全局提示

全局展示操作反馈信息。

## 基本用法

使用 `Message.info` | `Message.success` | `Message.warning` | `Message.error` 来显示对应类型的全局提示。

```tsx
import React from 'react';
import { Button, Message } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button
          type="primary"
          onClick={() => {
            Message.info("这是一条默认提示");
          }}
        >
          默认提示
        </Button>
        <Button
          type="success"
          onClick={() => {
            Message.success("这是一条成功提示");
          }}
        >
          成功提示
        </Button>
        <Button
          type="warning"
          onClick={() => {
            Message.warning("这是一条警告提示");
          }}
        >
          警告提示
        </Button>
        <Button
          type="error"
          onClick={() => {
            Message.error("这是一条错误提示");
          }}
        >
          错误提示
        </Button>
      </div>
    </div>
  );
};

export default App;
```

## 设置提示持续时间

可通过第二个参数来设置提示持续的时间，默认为 `3`，当设为 `0` 时保持提示不关闭。

```tsx
import React from 'react';
import { Button, Message } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button
          onClick={() => {
            Message.info("这是一条不关闭的提示", 0);
          }}
        >
          提示不关闭
        </Button>
        <Button
          onClick={() => {
            Message.info("这是一条持续 2 秒的提示", 2);
          }}
        >
          提示 2 秒后关闭
        </Button>
        <Button
          onClick={() => {
            Message.info("这是一条持续 10 秒的提示", 10);
          }}
        >
          提示 10 秒后关闭
        </Button>
      </div>
    </div>
  );
};

export default App;
```

## 设置提示展示位置

可通过第三个参数来设置提示出现的位置，默认为 `top`。

```tsx
import React from 'react';
import { Button, Message } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button
          onClick={() => {
            Message.info("这是一条不关闭的提示", 3);
          }}
        >
          展示在顶部的提示
        </Button>
        <Button
          onClick={() => {
            Message.info("这是一条不关闭的提示", 3, "bottom");
          }}
        >
          展示在底部的提示
        </Button>
      </div>
    </div>
  );
};

export default App;
```

## 手动关闭所有提示

可通过 `Message.closeAll()` 来关闭所有提示。

```tsx
import React from 'react';
import { Button, Message } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button
          onClick={() => {
            Message.info("这是一条不关闭的提示", 0);
          }}
        >
          提示不关闭
        </Button>
        <Button
          onClick={() => {
            Message.info("这是一条持续 2 秒的提示", 2);
          }}
        >
          提示 2 秒后关闭
        </Button>
        <Button
          onClick={() => {
            Message.closeAll();
          }}
        >
          关闭所有提示
        </Button>
      </div>
    </div>
  );
};

export default App;
```

## 传入配置的用法

除上述方式外，也可以传入 `{ content: "这是一条成功提示", duration: 5000 }` 对象来使用全局提示。

```tsx
import React from 'react';
import { Button, Message } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button
          onClick={() => {
            Message.info({
              content: "这是一条默认提示"
            });
          }}
        >
          默认提示
        </Button>
        <Button
          onClick={() => {
            Message.info({
              content: "这是一条持续 10 秒的提示",
              duration: 10,
              placement: 'bottom'
            });
          }}
        >
          提示 10 秒后关闭
        </Button>
      </div>
    </div>
  );
};

export default App;
```

<API />
