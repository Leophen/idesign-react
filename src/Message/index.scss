.i-message {
  display: flex;
  align-items: center;
  box-sizing: border-box;
  width: fit-content;
  padding: 10px 16px;
  color: var(--i-font-2);
  font-size: 14px;
  background-color: var(--i-bg);
  border-radius: 6px;
  outline: 0;
  box-shadow: var(--i-shadow);
  transform: translateX(0);
  .i-icon {
    margin-right: 8px;
  }
  .icon-TipInfo {
    color: var(--i-primary);
  }
  .icon-TipCheckFill {
    color: var(--i-success);
  }
  .icon-TipWarningFill {
    color: var(--i-warning);
  }
  .icon-TipCloseFill {
    color: var(--i-error);
  }
}

.i-message-wrapper {
  position: fixed;
  top: 20px;
  left: 50%;
  z-index: 5000;
  display: flex;
  flex-direction: column;
  align-items: center;
  transform: translateX(-50%);
  .i-message:not(:last-of-type) {
    margin-bottom: 16px;
  }
  &.i-message-wrapper__bottom {
    top: unset;
    bottom: 20px;
  }
}
