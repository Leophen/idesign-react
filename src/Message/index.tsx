import React from 'react';
import ReactDOM from 'react-dom';
import './index.scss';
import Icon from '../Icon';
import { TransitionGroup } from 'react-transition-group'
import Transition from '../Transition'
import { MessageConfig, MessageProps } from './type';

const Message: React.FC<MessageProps> & {
  success?: any
  warning?: any
  error?: any
  info?: any
  closeAll?: any
} = (props) => {
  const {
    type = 'info',
    content = ''
  } = props;

  return (
    <div
      className='i-message'
    >
      <Icon
        name={
          {
            info: 'TipInfo',
            success: 'TipCheckFill',
            warning: 'TipWarningFill',
            error: 'TipCloseFill',
          }[type]
        }
        size={20}
      />
      {content}
    </div>
  );
};

interface MessageContainerProps {
  messageListData: any[]
  placement?: 'top' | 'bottom';
}

const MessageContainer: React.FC<MessageContainerProps> = (props) => {
  const { messageListData, placement = 'top' } = props
  return (
    <TransitionGroup>
      {
        messageListData.map(({ key, type, content }) => (
          <Transition
            timeout={200}
            in
            animation={`slide-in-${placement}`}
            key={key}
          >
            <Message type={type} content={content} />
          </Transition>
        ))
      }
    </TransitionGroup>
  )
}

const messageList: any = {
  'top': [],
  'bottom': []
}

const createMessageWrapper = (placement: 'top' | 'bottom') => {
  const idName = `i-message-wrapper__${placement}`
  let el = document.querySelector(`#${idName}`)
  if (!el) {
    el = document.createElement('div')
    el.className = `i-message-wrapper ${idName}`
    el.id = idName
    document.body.append(el)
  }
  ReactDOM.render(<MessageContainer placement={placement} messageListData={messageList[placement]} />, el)
}

createMessageWrapper('top')
createMessageWrapper('bottom')

const updateMessageContainer = (config: any, mode = 'add') => {
  const location = config.placement
  let el = document.querySelector(`#i-message-wrapper__${location}`)
  if (mode === 'add') { // 添加模式
    location === 'top' ? messageList[location].push(config) : messageList[location].unshift(config)
    // 延迟更新 DOM
    if (config.duration > 0) {
      setTimeout(() => {
        messageList[location].map((item: any, index: number) => {
          if (item.key === config.key) {
            messageList[location].splice(index, 1)
          }
        })
        ReactDOM.render(<MessageContainer placement={location} messageListData={messageList[location]} />, el)
      }, config.duration * 1000)
    }
  } else { // 删除模式
    messageList[location] = []
  }
  ReactDOM.render(<MessageContainer placement={location} messageListData={messageList[location]} />, el)
}

const openMessage = (
  type: 'info' | 'success' | 'warning' | 'error',
  content: string | MessageConfig,
  duration = 3,
  placement = 'top'
) => {
  updateMessageContainer({
    key: Date.now(),
    type,
    content: typeof content === 'object' ? content?.content : content,
    duration: typeof content === 'object' ? content?.duration || 3 : duration,
    placement: typeof content === 'object' ? content?.placement || 'top' : placement
  }, 'add')
}

const closeMessage = (
  content: string | MessageConfig,
  placement = 'top'
) => {
  updateMessageContainer({
    placement: typeof content === 'object' ? content?.placement || 'top' : placement
  }, 'del')
}

Message.info = (content: string | MessageConfig, duration?: number, placement?: 'top' | 'bottom') => openMessage('info', content, duration, placement)
Message.success = (content: string | MessageConfig, duration?: number, placement?: 'top' | 'bottom') => openMessage('success', content, duration, placement)
Message.warning = (content: string | MessageConfig, duration?: number, placement?: 'top' | 'bottom') => openMessage('warning', content, duration, placement)
Message.error = (content: string | MessageConfig, duration?: number, placement?: 'top' | 'bottom') => openMessage('error', content, duration, placement)
Message.closeAll = (content: string | MessageConfig, duration?: number, placement?: 'top' | 'bottom') => closeMessage(content, placement)

Message.displayName = 'Message';

export default Message;
