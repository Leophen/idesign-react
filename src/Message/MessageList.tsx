import React, { useState } from 'react';
import Message from './Message';
import { TransitionGroup } from 'react-transition-group';
import Transition from '../Transition';
import { MessageListProps } from './type';
import './index.scss';

const MessageList: React.FC<MessageListProps> = (props) => {
  const { listData, position = 'top' } = props;
  const [enteredIds, setEnteredIds] = useState<number[]>([])

  const handleEntered = (id: number) => {
    enteredIds.push(id)
    setEnteredIds([...enteredIds])
  }

  return (
    <TransitionGroup>
      {listData.map(({ id, type, content }) => (
        <Transition
          timeout={300}
          in
          animation={`message-${position}`}
          key={id}
          onEntered={() => handleEntered(id)}
        >
          <Message type={type} content={content} entered={enteredIds.includes(id)} />
        </Transition>
      ))}
    </TransitionGroup>
  );
};

MessageList.displayName = 'MessageList';

export default MessageList;
