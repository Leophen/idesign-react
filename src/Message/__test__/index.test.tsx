import { render } from '@testing-library/react';
import Message from '../Message';

describe('Message 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(<Message>消息提示</Message>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('type', () => {
    const { container } = render(<Message type="success">成功提示</Message>);
    expect(container.firstChild.firstChild).toHaveClass('icon-TipCheckFill');
  });
});



