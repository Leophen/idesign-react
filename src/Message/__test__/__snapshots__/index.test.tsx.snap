// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-message"
  >
    <i
      class="i-icon icon-TipInfo"
      style="font-size: 20px;"
    />
  </div>
</DocumentFragment>
`;
